### PR TITLE
Bug 1668901 - Add keyboard shortcut 'q' to switch between classified …

### PR DESCRIPTION
**Pull request by rinivelkum**
…and unclassified, and only unclassified tasks

### CODE REVIEW

**Comment by sarah-clements** [comment]
Hi, thanks for your contribution but please in the future do not start working on a bugzilla bug until you have had it assigned to you, per our contributing guidelines: https://github.com/mozilla/treeherder#contributing

Our team had this bug marked as work in a sprint for this week, but luckily no one had started working on it yet, so this is why we have the guidelines in place.

**Comment by rinivelkum** [comment]
Hi Sarah, sorry about not following the guidelines!
Should I proceed with the bugzilla information request?
Or close the pull request if it is already assigned?

**Comment by sarah-clements** [comment]
> Hi Sarah, sorry about not following the guidelines!
> Should I proceed with the bugzilla information request?
> Or close the pull request if it is already assigned?

 I'll assign you to the bug and review this pr shortly, just be sure to ask to be assigned to a bug next time :)

**Comment by rinivelkum** [comment]
Thank you, my username is "Rinivelkum".

File: ui/shared/ShortcutTable.jsx [hunk 1]
```diff
@@ -54,6 +54,15 @@ const ShortcutTable = function ShortcutTable() {
               </td>
               <td>Open the logviewer for the selected job</td>
             </tr>
+            <tr>
+              <td>
+                <kbd>q</kbd>
+              </td>
+              <td>
+                Toggle showing only classified, unclassified, retried and
+                usercancelled jobs
```
Comment (sarah-clements) [inline ui/shared/ShortcutTable.jsx:63]: I'd reword this as: `Toggle between unclassified failures and all failures (including retried and user cancelled jobs)`

**Comment by sarah-clements** [comment]
This looks great so far. The only thing it'd be helpful to add is to change the filters menu so that when a user hits this keyboard shortcut, the logic corresponds correctly to the filters menu. For example, see the filters menu below:

Screen Shot 2020-11-30 at 3 56 38 PM

The logic for that menu is different from the keyboard shortcut because currently when you deselect "classified" it shows unclassified AND all other jobs, which is not how the sheriffs want it. So "unclassified" and "classified" options should be updated to "all jobs", "all Failures" and "unclassified Failures". Then when the keyboard shortcut is used the checkbox on the menu will have "unclassified Failures" selected, otherwise "all Failures" selected (and the filters menu logic will work the same as the keyboard shortcut). Could you add this change to this pr please? :)

You can find the file here: `ui/job-view/headerbars/FiltersMenu.jsx` and the filter selection is probably updated by the query params that were set in the filtermodel.

**Comment by rinivelkum** [comment]
* Summary of my changes:
- I added the "All jobs", "All failures" and "Unclassified failures" Dropdown Items in place of "classified" and "unclassified".
- Imported two constants from the filter helpers file onto the FiltersMenu file for me to use.
- Modified the "q" shortcut description as requested.
- Added a "resultStatus" check for <'testfailed', 'busted', 'exception', 'retry', 'usercancel'> in the "isClassifiedFailures" function from FilterModel because when switching from "All jobs" to "All failures", "Unclassified failures" would get selected instead (and "q" or "All failures" would need pressing twice to switch to classified and failures").
That is because before, "isClassifiedFailures" would only check for the "classifiedState" parameter, and "All jobs" view has that one toggled.
- Switched from using "toggleClassifiedFilter()" on classified failures to using "toggleUnclassifiedFailures()".
By using the first we would get cases in which only classified failures would show when clicking "Unclassified failures".
I don't know how useful that is, but it can be reverted easily if needed.

* Logic of the buttons:
All jobs - switches from showing all jobs to showing all failures.
All failures - switches from showing testfailure, retried, usercancelled, busted, exception; both classified and unclassified,
to showing only unclassified.
Unclassified failures - switches from showing testfailure, retried, usercancelled, busted, exception; only unclassified,
to showing All jobs.

**Comment by sarah-clements** [comment]
> **Logic of the buttons:**
> 
>  * All jobs - switches from showing all jobs to showing all failures.
> 
>  * All failures - switches from showing testfailure, retried, usercancelled, busted, exception; both classified and unclassified,
>       to showing only unclassified.
> 
>  * Unclassified failures - switches from showing testfailure, retried, usercancelled, busted, exception; only unclassified,
>       to showing All jobs.

Thanks for the detailed summary. I think what would be clearer is to use OR logic and for only one of those three options to be selected (have a check mark) at a time. If "All jobs" is selected, "All Failures" and "Unclassified Failures" shouldn't be since its already implied that they are shown. And if "All Failures" is selected, then "Unclassified failures" shouldn't because that is also implied (same for when you hit the q button the first time, only "All Failures" should be shown, then only "Unclassified Failures" when only those are shown/toggled.)

Also, please add your changes in separate commits, prepended with something like "changes based on feedback". Makes it easier to review (and they'll be squashed when this is merged). :)

**Comment by rinivelkum** [comment]
Changes:
- Modified the condition for the checkmark to be selected, now they are shown one at a time when the view is active.

Hope I understood correctly your feedback about only one option being selected at a time, the only downside I can see about the changes is that once you modify manually one of the filters, the checkmark is hidden, since it doesn't correspond to the option parameters, and we get a menu with no toggled view.

This could be fixed by setting more flexible conditions for the checkmark, but not flexible enough that they overlap and get toggled simultaneously (my idea would be to make "All failures" remain toggled even when other filters such as 'running',  'pending', 'runnable' are active; but not 'success', so that it doesn't overlap with "All jobs".
Applying the same logic to the "Unclassified failures" option, this one could remain toggled the same way as "All failures", excluding the Classified failures, or it could remain specific in the same way as "All jobs" is, and toggle strictly when only unclassified failures are being shown.

**Comment by sarah-clements** [comment]
> Changes:
> 
> * Modified the condition for the checkmark to be selected, now they are shown one at a time when the view is active.
> 
> 
> Hope I understood correctly your feedback about only one option being selected at a time, the only downside I can see about the changes is that once you modify manually one of the filters, the checkmark is hidden, since it doesn't correspond to the option parameters, and we get a menu with no toggled view.

I don't see that as a downside as long as the filters work accordingly :) This would be a great thing to test with the sheriffs, who would primarily be affected by this, to see if the change in behavior is more intuitive or less so. I'll see if I can find a time today to deploy your pr to our development instance and ask one or two of the sheriffs to leave feedback here.

**Comment by sarah-clements** [comment]
I've got some feedback and there is a request to add "Classified Failures" back into the Filters menu (so only classified failures, but not changing the keyboard shortcut logic at all). Also I got good feedback on how changing a filter selection changes whether one of the "All Jobs", "Classified Failures" and "All Failures" options is shown and it was clear so no need to change it :) 

One little thing that needs to be fixed is this red dot to the left of the Filters menu (see `filter-chicklets` in the code base). It should be deselected, (not completely red) if `testfailed` and `busted` are deselected. Right now that only seems to happen if `retry` is also selected:

729224f814bbd75570dc5e7bb298d3dc392a6b27

Just these few changes and this is ready to merge!

**Comment by sarah-clements** [comment]
Also, I forgot to mention please run `yarn lint --fix` and `yarn tests` locally with your next commit to ensure all tests pass. I can see the linting check is failing (if you click on the details of the red x above).

**Comment by rinivelkum** [comment]
Glad the changes are welcomed!
Sarah, could you please confirm this?
- When pressing the red (failures) chicklet the FilterMenu and the view show "testfailed", "busted" and "exception" being toggled, not retry. Deselecting "testfailed" and "busted" indeed does not hide the failure chicklet, but that would be because "exception" is still toggled, not "retry", since that one belongs to "thFilterGroups.nonfailures".

Also two questions regarding the "classified failures" button being brought back to the FilterMenu: 
- Should it's positioning be next to "Unclassified failures", or under the divider, next to "Superseeded Only"?
- Should the behavior be the same as before? Or should it switch to a view containing strictly classified failures?

**Comment by sarah-clements** [comment]
> Glad the changes are welcomed!
> Sarah, could you please confirm this?
> 
>    * When pressing the red (failures) chicklet the FilterMenu and the view show "testfailed", "busted" and "exception" being toggled, not "retry". Deselecting "testfailed" and "busted" indeed does not hide the failure chicklet, but that would be because "exception" is still toggled, not "retry", since that one belongs to "thFilterGroups.nonfailures".
> 
Ah, you're right. Let's leave that as-is.

> Also two questions regarding the "classified failures" button being brought back to the FilterMenu:
> 
>  * Should it's positioning be next to "Unclassified failures", or under the divider, next to "Superseeded Only"?
> 
>  * Should the behavior be the same as before? Or should it switch to a view containing strictly classified failures?

It should go before or after "Unclassified failures" and only show classified failures when selected.

**Comment by rinivelkum** [comment]
Summary of my changes:
- Added back the "Classified Failures" button, now strictly displaying classified failures.
- Repurposed the "toggleClassifiedFailures" function to display either only "Classified Failures" or "All failures", depending on the value of the boolean "showUnclassified" parameter, hope this is intuitive to use.
- Repurposed the "isClassifiedFailures" function to check if we are both in a "Classified Failures" view or in an "All failures" view.
- Added a condition on the "All jobs" button that checks if we are showing both classified and unclassified failures, since not adding that condition would lead to "All jobs" being checked when selecting "Classified Failures" and deselecting all the chicklets, which in turn would show all the classified jobs (including success, running...)  and toggle the chicklet.
- Kept the "Classified Failures" view switch the same as "Unclassified Failures", meaning that when pressed twice, you are shown the "All jobs" view.

(Also got another lint failure which I'm fixing atm)

**Comment by sarah-clements** [comment]
> **Summary of my changes**:
> 
>     * Added back the "Classified Failures" button, now strictly displaying classified failures.
> 
>     * Repurposed the "toggleClassifiedFailures" function to display either only "Classified Failures" or "All failures", depending on the value of the boolean "showUnclassified" parameter, hope this is intuitive to use.
> 
>     * Repurposed the "isClassifiedFailures" function to check if we are both in a "Classified Failures" view or in an "All failures" view.
> 
>     * Added a condition on the "All jobs" button that checks if we are showing both classified and unclassified failures, since not adding that condition would lead to "All jobs" being checked when selecting "Classified Failures" and deselecting all the chicklets, which in turn would show all the classified jobs (including success, running...)  and toggle the chicklet.
> 
>     * Kept the "Classified Failures" view switch the same as "Unclassified Failures", meaning that when pressed twice, you are shown the "All jobs" view.
> 
> 
> (Also got another lint failure which I'm fixing atm)

Thanks! I'll give this a review and test tomorrow.

**Review by sarah-clements — APPROVED** [review]
This looks great - thanks for all your work on this! :)
